src/Datasource/EntityTrait.php

```diff
@@ -242,13 +243,39 @@ public function set($property, $value = null, array $options = [])
         $options += ['setter' => true, 'guard' => $guard];
 
         foreach ($property as $p => $value) {
+            $nested = false;
+            if (strpos($p, '.') > 0 && !array_key_exists($p, $this->_properties)) {
+                list($p, $nested) = explode('.', $p, 2);
+            }
+
             if ($options['guard'] === true && !$this->accessible($p)) {
                 continue;
             }
 
             $this->dirty($p, true);
 
-            if (!array_key_exists($p, $this->_original) &&
+            if ($nested) {
+                if (!array_key_exists($p, $this->_properties)
+                    || is_null($this->_properties[$p])
+                ) {
+                    $value = new \Cake\ORM\Entity([$nested => $value]);
```

Comment:
I'm starting to think that throwing an Exception in `::getOriginal()` when the requested path is dotted but ::_original data is not available all the way down the chain is the best course. Basically saying, "don't try that, because we can't answer your question in this case." There may be side-effects that I'm not anticipating with that yet, but lemme poke at the tests here a bit.